internal/graph/check.go

```diff
@@ -0,0 +1,536 @@
+)
+
+type setOperatorType int
+
+const (
+	unionSetOperator setOperatorType = iota
+	intersectionSetOperator
+	exclusionSetOperator
+)
+
+type checkOutcome struct {
+	resp *openfgapb.CheckResponse
+	err  error
+}
+
+// LocalChecker implements Check in a highly concurrent and localized manner. The
+// Check resolution is limited per branch of evaluation by the concurrencyLimit.
+type LocalChecker struct {
+	ds               storage.RelationshipTupleReader
+	dispatcher       dispatcher.CheckDispatcher
+	concurrencyLimit uint32
+}
+
+// NewLocalChecker constructs a LocalChecker that can be used to evaluate a Check
+// request locally and with a high degree of concurrency.
+func NewLocalChecker(
+	ds storage.RelationshipTupleReader,
+	concurrencyLimit uint32,
+) *LocalChecker {
+	checker := &LocalChecker{ds: ds, concurrencyLimit: concurrencyLimit}
+	checker.dispatcher = checker // todo: replace with a different CheckDispatcher once we support remote dispatching
+
+	return checker
+}
+
+// CheckHandlerFunc defines a function that evaluates a CheckResponse or returns an error
+// otherwise.
+type CheckHandlerFunc func(ctx context.Context) (*openfgapb.CheckResponse, error)
+
+// CheckFuncReducer defines a function that combines or reduces one or more CheckHandlerFunc into
+// a single CheckResponse with a maximum limit on the number of concurrent evaluations that can be
+// in flight at any given time.
+type CheckFuncReducer func(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandlerFunc) (*openfgapb.CheckResponse, error)
+
+// resolver concurrently resolves one or more CheckHandlerFunc and yields the results on the provided resultChan.
+// Callers of the 'resolver' function should be sure to invoke the callback returned from this function to ensure
+// every concurrent check is evaluated. The concurrencyLimit can be set to provide a maximum number of concurrent
+// evaluations in flight at any point.
+func resolver(ctx context.Context, concurrencyLimit uint32, resultChan chan<- checkOutcome, handlers ...CheckHandlerFunc) func() {
+	limiter := make(chan struct{}, concurrencyLimit)
+
+	var wg sync.WaitGroup
+
+	checker := func(fn CheckHandlerFunc) {
+		resp, err := fn(ctx)
+		resultChan <- checkOutcome{resp, err}
+		<-limiter
+		wg.Done()
+	}
+
+	wg.Add(1)
+	go func() {
+	outer:
+		for _, handler := range handlers {
+			fn := handler // capture loop var
+
+			select {
+			case limiter <- struct{}{}:
+				wg.Add(1)
+				go checker(fn)
+			case <-ctx.Done():
+				break outer
+			}
+		}
+
+		wg.Done()
+	}()
+
+	return func() {
+		wg.Wait()
+		close(limiter)
+	}
+}
+
+// union implements a CheckFuncReducer that requires any of the provided CheckHandlerFunc to resolve
+// to an allowed outcome. The first allowed outcome causes premature termination of the reducer.
+func union(ctx context.Context, concurrencyLimit uint32, handlers ...CheckHandlerFunc) (*openfgapb.CheckResponse, error) {
+
+	cctx, cancel := context.WithCancel(ctx)
```

Comment:
Any reason to create a new variable for the context?

Reply:
Not really. I renamed to override `ctx`.